sympy/tensor/array/array_comprehension.py

```diff
@@ -103,18 +107,88 @@ def free_symbols(self):
                 expr_free_sym = expr_free_sym.union(sup.free_symbols)
         return expr_free_sym
 
+    @property
+    def shape(self):
+        """
+        Returns the shape of the expanded array, which can have symbols. Note that both
+        the lower and the upper bounds are included while calculating the shape.
+
+        Examples
+        ========
+
+        >>> from sympy.tensor.array import ArrayComprehension
+        >>> from sympy.abc import i, j, k
+        >>> a = ArrayComprehension(10*i + j, (i, 1, 4), (j, 1, 3))
+        >>> a.shape
+        (4, 3)
+        >>> b = ArrayComprehension(10*i + j, (i, 1, 4), (j, 1, k+3))
+        >>> b.shape
+        (4, k + 3)
+        """
+        return self._shape
+
+    def rank(self):
+        """
+        Returns the rank of the expanded array.
+
+        Examples
+        ========
+
+        >>> from sympy.tensor.array import ArrayComprehension
+        >>> from sympy.abc import i, j, k
+        >>> a = ArrayComprehension(10*i + j, (i, 1, 4), (j, 1, 3))
+        >>> a.rank()
+        2
+        """
+        return self._rank
+
+    def __len__(self):
+        """
+        Overload common function len().Returns the number of element in the expanded
+        array. Note that symbolic length is not supported and will raise an error.
+
+        Examples
+        ========
+
+        >>> from sympy.tensor.array import ArrayComprehension
+        >>> from sympy.abc import i, j, k
+        >>> a = ArrayComprehension(10*i + j, (i, 1, 4), (j, 1, 3))
+        >>> len(a)
+        12
+        """
+        if len(self._loop_size.free_symbols) != 0:
+            raise ValueError('Symbolic length is not supported')
```

Comment:
Is it possible to find a bypass so that we can return an `expression` rather than `integer` for `len()`?
Or maybe another function/property like `size` to return this length of the expanded array.

Reply:
AFAIK, `__len__` always returns an integer, so recommended is using `size`. Something similar has been done in `Range` in `sympy.sets.fancysets`.

Reply:
what about 
```python
return int(product([sup-inf for var, inf, sup in self.limits]))
```
?

Reply:
I am sorry but I don't really understand how product will help in this case. Do you mean `itertool.product`? In fact, the problem is when we have an array like
```
>>> a = ArrayComprehension(10*i + j, (i, 1, 4), (j, 1, k + 3))
>>> len(a)
4 * (k + 3)
```
it is supposed to have an expression, but len() accepts only integer as return type.
For your information, a bound like `(i, j, j+4)` will return 4 as length so there is not problem if the difference is an `Integer`.